impl/src/main/java/org/jboss/weld/bootstrap/Tracker.java

```diff
@@ -0,0 +1,142 @@
+ * by the @authors tag. See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.weld.bootstrap;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.concurrent.TimeUnit;
+
+import org.jboss.weld.logging.BootstrapLogger;
+
+/**
+ * A simple tracker used to monitor bootstrap operations. It is not thread-safe and may not be shared between threads.
+ *
+ * @author Martin Kouba
+ */
+interface Tracker extends AutoCloseable {
+
+    String OP_START_CONTAINER = "startContainer";
+    String OP_INIT_SERVICES = "initServices";
+    String OP_CONTEXTS = "builtinContexts";
+    String OP_READ_DEPLOYMENT = "readDeploymentStructure";
+    String OP_START_INIT = "startInitialization";
+    String OP_DEPLOY_BEANS = "deployBeans";
+    String OP_VALIDATE_BEANS = "validateBeans";
+    String OP_END_INIT = "endInitialization";
+    String OP_BBD = "BeforeBeanDiscovery";
+    String OP_ATD = "AfterTypeDiscovery";
+    String OP_ABD = "AfterBeanDiscovery";
+    String OP_ADV = "AfterDeploymentValidation";
+
+    /**
+     * Starts an operation.
+     *
+     * @param operation
+     * @return self
+     */
+    default Tracker start(String operation) {
+        return this;
+    }
+
+    /**
+     * End the last operation on the stack.
+     *
+     * @return self
+     */
+    default Tracker end() {
+        return this;
+    }
+
+    default void split(String info) {
+    }
+
+    /**
+     * End all operations.
+     */
+    default void close() {
+    }
+
+    static Tracker NOOP_INSTANCE = new Tracker() {
+    };
+
+    static Tracker create(String startOperation) {
+        return create().start(startOperation);
+    }
+
+    static Tracker create() {
+        return BootstrapLogger.TRACKER_LOG.isDebugEnabled() ? new LoggingTracker() : NOOP_INSTANCE;
+    }
+
+    class LoggingTracker implements Tracker {
+
+        private final List<Operation> operations;
+
+        private LoggingTracker() {
+            this.operations = new LinkedList<>();
+        }
+
+        @Override
+        public Tracker start(String operation) {
+            if (!operations.isEmpty()) {
+                operation = operations.get(operations.size() - 1).name + " > " + operation;
+            }
+            operations.add(new Operation(operation));
+            BootstrapLogger.TRACKER_LOG.debugf("ST: %s ", operation);
```

Comment:
Yep, ST stands for START and EN stands for END. The intention was to use abbreviations so that it's easy to align the log messages. Otherwise we would have to add gaps, something like...
```
[main] DEBUG org.jboss.weld.BootstrapTracker - START: endInitialization 
[main] DEBUG org.jboss.weld.BootstrapTracker - END:   endInitialization (6 ms)
```